cylc/flow/ws_data_mgr.py

```diff
@@ -0,0 +1,393 @@
+#!/usr/bin/env python3
+
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
+# Copyright (C) 2008-2019 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Manage the data feed for the User Interface Server."""
+
+from time import time
+from copy import copy
+import ast
+
+from cylc.flow.task_id import TaskID
+from cylc.flow.suite_status import (
+    SUITE_STATUS_HELD, SUITE_STATUS_STOPPING,
+    SUITE_STATUS_RUNNING, SUITE_STATUS_RUNNING_TO_STOP,
+    SUITE_STATUS_RUNNING_TO_HOLD)
+from cylc.flow.task_state_prop import extract_group_state
+from cylc.flow.wallclock import (
+    TIME_ZONE_LOCAL_INFO, TIME_ZONE_UTC_INFO, get_utc_mode)
+from cylc.flow.task_job_logs import JOB_LOG_OPTS
+from cylc.flow import __version__ as CYLC_VERSION
+from cylc.flow.ws_messages_pb2 import (
+    PbFamily, PbFamilyProxy, PbTask, PbTaskProxy, PbWorkflow,
+    PbEdge, PbEdges, PbEntireWorkflow)
+
+
+class WsDataMgr(object):
+    """Manage the data feed for the User Interface Server."""
+
+    TIME_FIELDS = ['submitted_time', 'started_time', 'finished_time']
+
+    def __init__(self, schd):
+        self.schd = schd
+        self.tasks = {}
+        self.task_proxies = {}
+        self.task_states = {}
+        self.families = {}
+        self.family_proxies = {}
+        self.state_count_totals = {}
+        self.state_count_cycles = {}
+        self.edges = {}
+        self.graph = PbEdges()
+        self.workflow_id = f"{self.schd.owner}/{self.schd.suite}"
+        self.workflow = PbWorkflow()
+
+    def initiate_data_model(self):
+        """Initiate or Update data model."""
+        update_time = time()
+        config = self.schd.config
+        tasks = {}
+        task_proxies = {}
+        families = {}
+        family_proxies = {}
+        edges = {}
+        graph = PbEdges()
+        workflow = PbWorkflow(
+            checksum=f"{self.workflow_id}@{update_time}",
+            id=self.workflow_id,
+        )
+
+        all_states = []
+
+        # Compute state_counts (total, and per cycle).
+        state_count_totals = {}
```

Comment:
Declared again around line ~ 257. Maybe we can remove this declaration, or that one?

Reply:
N/A